pkg/icingadb/history/retention.go

```diff
@@ -172,6 +173,20 @@ func (r *Retention) Start(ctx context.Context) error {
 			r.logger.Debugf("Skipping history retention for category %s", stmt.Category)
 			continue
 		}
+		// #nosec G115 -- First if-clause checks int overflow
+		if days > math.MaxInt || time.Now().AddDate(0, 0, -int(days)).Before(time.Unix(0, 0)) {
+			// Ensure that the time cannot overflow.
+			//
+			// First, make sure that days do not exceed max int, as otherwise casts will overflow and later subtractions
+			// may overflow as well. By being capped to [0, max int), for each current timestamp t, t-days > min int.
+			// As the database schema uses an unsigned timestamp, the second check ensures that the relative offset,
+			// later called olderThan, does not overflow.
+			//
+			// For the periodic.Start function below, time marches forward while the value of days is static.
+			// Subtracting days from an increasing timestamp will also not result in a negative or overflowing value.
+			r.logger.Errorf("Skipping history retention for category %s as %d days will overflow", stmt.Category, days)
+			continue
+		}
```

Comment:
> `days > math.MaxInt`

That still allows for `days` = $2^{63} - 1$ in 64 bit builds which is too much to handle for `AddDate()`.

> By being capped to [0, max int), for each current timestamp t, **t-days > min int**.

How is min int relevant here? Wouldn't be the min `time.Time` be the relevant thing here (or maybe the internals of `AddDate()` if that breaks even earlier)?

From https://github.com/Icinga/icingadb/pull/821#discussion_r1792140287:

> reducing days to uint16 sounds good to me too. Though I would do that in a separate PR.

Though doing it in a separate PR sounds like this would either be a prerequisite of this PR or this PR has to do a similar bounds check. Doing another PR first doesn't sound like a bad idea for me from a changelog perspective: a separate PR for "limit history retention to $2^16$ days (~180 years)" might be more helpful than hiding it in a linter PR.

Reply:
The comment contained "outdated" fractions and implementation knowledge of `time.Time`. I have cleaned that up.

Furthermore, I have added a third check[^1] to check if the calculated time with `AddDate()` is after the current date, implying an overflow in the `AddDate()` internals.

However, I can also change the type to `uint16` and will prepare another PR for that. This will still require the current time comparison check against Unix epoch as subtraction `math.MaxUint16` days will currently end up in [1845](https://go.dev/play/p/7oXTiOTZ3Nn).

[^1]: yay, more checks

Reply:
The type change was proposed in #824.

Reply:
> The type change was proposed in #824.

And it was just merged, which should actually make the `MaxInt` check and this new check already obsolete:

> Furthermore, I have added a third check[1](#user-content-fn-1-bb8cabe0a4b2ff53552d596385fb8e0b) to check if the calculated time with `AddDate()` is after the current date, implying an overflow in the `AddDate()` internals.

Reply:
Now this code is way easier to read :)

Reply:
> Though doing it in a separate PR sounds like this would either be a prerequisite of this PR or [...]

Well, there was a reason why I said that earlier :sweat_smile: